contracts/cw721-balance-voting/src/contract.rs

```diff
@@ -0,0 +1,94 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    to_binary, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult, Uint128,
+};
+use cw2::set_contract_version;
+
+use crate::error::ContractError;
+use crate::msg::{ExecuteMsg, InstantiateMsg, QueryMsg, TokenInfo};
+use crate::state::TOKEN;
+
+const CONTRACT_NAME: &str = "crates.io:cw721-balance-voting";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    match msg.token_info {
+        TokenInfo::Existing { address } => {
+            let address = deps.api.addr_validate(&address)?;
+            TOKEN.save(deps.storage, &address)?;
+            Ok(Response::default()
+                .add_attribute("action", "instantiate")
+                .add_attribute("token", "existing_nft")
+                .add_attribute("token_address", address))
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    _deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {}
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::TokenContract {} => query_token_contract(deps),
+        QueryMsg::VotingPowerAtHeight { address, height: _ } => {
+            query_voting_power_at_height(deps, env, address)
+        }
+        QueryMsg::TotalPowerAtHeight { height: _ } => query_total_power_at_height(deps, env),
+        QueryMsg::Info {} => query_info(deps),
+    }
+}
+
+pub fn query_token_contract(deps: Deps) -> StdResult<Binary> {
+    let token = TOKEN.load(deps.storage)?;
+    to_binary(&token)
+}
+
+pub fn query_voting_power_at_height(deps: Deps, env: Env, address: String) -> StdResult<Binary> {
+    let token = TOKEN.load(deps.storage)?;
+    let address = deps.api.addr_validate(&address)?;
+    let balance: cw721::TokensResponse = deps.querier.query_wasm_smart(
```

Comment:
I think we'd also need some sort of staking contract for cw721s. stake-cw20 could be used for inspiration :) The reason we require staking before voting in the cw20 case is to ensure that someone can't vote then send their tokens to another wallet and vote again, and also so that the proposer knows who has "registered to vote" for their proposal. you can thinking of staking as an alias for "registering your tokens / NFTs to vote"

Reply:
This is implemented as a separate contract in #309